File: lib/features/profile/controller/profile_controller.dart
import 'package:flutter/material.dart';
import 'package:protfolio_app/core/helpers/error_handler.dart';
import 'package:protfolio_app/features/profile/model/profile.dart';
import 'package:protfolio_app/features/profile/service/profile_service.dart';

class ProfileController extends ChangeNotifier {
  //step1 ----> here I will Create the class's variables :
  final ProfileService _service = ProfileService();
  Profile? _profile;
  bool _isLoading = false;
  String? _errorMessage;

  //step2 ----> here I will Create a getter && setter  :
  Profile? get profile => _profile;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  //step3 ----> here I will Create a function to load profile  :
  Future<void> loadProfile(int id) async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();
      _profile = await _service.getProfileById(id);
    } catch (e) {
      _errorMessage = ErrorHandler.getMessage(e);
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step4 ----> here I will Create a function to add profile :
  Future<void> addProfile(Profile profile) async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();
      _profile = await _service.addProfile(profile);
    } catch (e) {
      _errorMessage = ErrorHandler.getMessage(e);
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step5 ----> here I will Create a function to update profile :
  Future<void> updateProfile(int id, Profile profile) async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();
      _profile = await _service.updateProfile(id, profile);
    } catch (e) {
      _errorMessage = ErrorHandler.getMessage(e);
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step6 ----> here I will Create a function to delete profile :
  Future<void> deleteProfile(int id) async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();
      await _service.deleteProfile(id);
      _profile = null;
    } catch (e) {
      _errorMessage = ErrorHandler.getMessage(e);
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}

File: Backend/src/main/java/com/apps/PortfolioApp/Controller/ProfileController.java
package com.apps.PortfolioApp.Controller;

import org.springframework.web.bind.annotation.*;
import com.apps.PortfolioApp.DTO.ProfileDTO;
import com.apps.PortfolioApp.Service.ProfileService;
import jakarta.validation.Valid;

@RestController
@RequestMapping("/profiles")
public class ProfileController {
    // Step1 ----> Create an object from profile Service class :
    private final ProfileService profileService;

    // Step2 ----> here I will Create a constructor :
    public ProfileController(ProfileService profileService) {
        this.profileService = profileService;
    }

    // step3 ----> create a function to save the profile to db:
    @PostMapping
    public ProfileDTO insert(@RequestBody @Valid ProfileDTO profileDTO) {
        return profileService.insertProfile(profileDTO);
    }

    // step5 ----> create a function to get the Specific profile by id from db:
    @GetMapping("/{id}")
    public ProfileDTO getById(@PathVariable int id) {
        return profileService.getProfileById(id);
    }

    // step7 ----> create a function to update the profile to db:
    @PutMapping("/{id}")
    public ProfileDTO update(@PathVariable int id, @RequestBody @Valid ProfileDTO profileDTO) {

        return profileService.updateProfile(id, profileDTO);
    }

    // step8----> create a function to delete the Specific profile by id from db:
    @DeleteMapping("/{id}")
    public void delete(@PathVariable int id) {
        profileService.deleteProfile(id);
    }
}
